nixos/modules/config/system-path.nix

```diff
@@ -7,6 +7,97 @@ with lib;
 
 let
 
+  # This type can accept three different types of expressions:
+  #   * A list of packages like `[ pkgs.foo pkgs.bar ]`
+  #   * An attribute set of packages like `{ inherit (pkgs) foo bar; }`
+  #   * A more structured attribute set of packages like `{ "foo".package = pkgs.foo.override { hasBar = false; }; }`
+  #
+  # The first two cases will be coerced into the third one, which allows the
+  # package to be disabled and its package and list of outputs to be overridden.
+  packagesType =
+    let
+      inherit (lib) types;
+
+      checkAttrsOfDrvs = attrs:
+        let
+          drvs = lib.attrValues attrs;
+        in
+        lib.isAttrs attrs &&
+        lib.length drvs > 0 &&
+        lib.all (drv: lib.isDerivation drv) drvs;
+
+      outputsFor = drv:
+        if drv.outputSpecified or false
+        then [ drv.outputName ]
+        else ((drv.meta or {}).outputsToInstall or [ (drv.outputName or "out") ]);
+
+      mkOutputs =
+        let
+          mkEntry = output: lib.nameValuePair output true;
+        in
+        outputs: lib.listToAttrs (map mkEntry outputs);
+
+      drvToAttrs = drv:
+        let
+          defaultOutput = drv.${(lib.head (drv.outputs or [ "out" ]))}
+            or (throw "something is up with the outputs of this derivation: ${drv}");
+        in
+        {
+          ${defaultOutput.name} = {
+            package = defaultOutput;
+            outputs = mkOutputs (outputsFor drv);
+          };
+        };
+      drvToAttrsFolder = acc: drv: lib.recursiveUpdate acc (drvToAttrs drv);
+
+      pkgsToAttrs = lib.foldl' drvToAttrsFolder {};
+    in
+    types.coercedTo
+      (types.listOf types.package)
+      pkgsToAttrs
+      (types.coercedTo
```

Comment:
It seems that this could be `attrsOf (coercedTo` instead. That would simplify the rest of the type and iiuc behave better for mixed definitions (e.g. `systemPackages = { foo = pkgs.foo; bar = { package = pkgs.bar }; }`, which seems like a good test case as well)

Reply:
I tried this, but the big downside is that we cannot control the names anymore in that case, and so we get a difference in behaviour between `environment.systemPackages = [ pkgs.hello ];` and `environment.systemPackages = { inherit (pkgs) hello; };`.

In the first case we get `"${pkgs.hello.name}"` as attribute name (since not every derivation has a pname), but in the second case we get `"hello"`, and so the way to override this package would depend on how it was originally put into `systemPackages`, which seems unnecessarily brittle.

So we need a naming that can be derived from the derivation since in the case of a list we don't have anything else to derive the name from when we convert the list into an attr set. And ideally we would then make sure that the two other ways of defining `systemPackages` use the exact same naming scheme, so that things are stable and predictable.

What do you think?

Reply:
I might have been unclear with my suggestion. I meant to construct the type as

```nix
coercedTo
  (listOf package)
  pkgsToAttrs
  (attrsOf t)
```

where `t` is the logic below, except treating a single package at a time.

The issue you're describing seems to be about `pkgsToAttrs`, which I didn't mean to touch.

Reply:
Regarding the `pkg.name` problem, I'd be happy to give them really bad names, such as the drv hash, so that overridability is only possible when the original definition has an explicit attribute name.

Suppose NixOS switches to `sudo-rs` by default, while using the list syntax. That would break all configs that have an override on `sudo`.
I'd say lists are bad, and explicit is good, so let's make people use the attrset syntax if they want to use this overriding feature.

Reply:
> I might have been unclear with my suggestion. I meant to construct the type as
> 
> ```nix
> coercedTo
>   (listOf package)
>   pkgsToAttrs
>   (attrsOf t)
> ```
> 
> where `t` is the logic below, except treating a single package at a time.
> 
> The issue you're describing seems to be about `pkgsToAttrs`, which I didn't mean to touch.

I think we're talking about the same thing, but what you're proposing doesn't really work well.

So what you're proposing is something like this:
```nix
coercedTo
  (listOf package)
  pkgsToAttrs
  (attrsOf (coercedTo
    package
    packageToStructuredAttrs
    (submodule { ... })
  ))
```

The problem is that in this case when we write
```nix
environment.systemPackages = { inherit (pkgs) hello; };
```
We get as attribute name `"hello"`, but when we convert from the list of packages, we get the full `name` attribute (or the drv hash or whatever we choose for it, but usually not the attribute name of the package in `pkgs`).
And so we cannot override between these two because the attribute names are inconsistent. And since our coercion function is only action on the values of the attribute sets, and not on the sets themselves, we cannot change the attribute names.

So in this case we cannot have the same attribute name in the final attribute set for
`systemPackages = [ pkgs.hello ];` and `systemPackages = { inherit (pkgs) hello; }`.

The big difference with the current implementation, is that when we have
```nix
coercedTo
  (listOf package)
  pkgsToAttrs
  (coercedTo
    (attrsOf package)
    packagesToStructuredAttrs
    (attrsOf (submodule { ... }))
  )
```
we can change the attribute names in `packagesToStructuredAttrs`, since we are converting one attrset into another.

Reply:
Agreed, but then we need to have the type that I'm currently proposing, as `{ inherit (pkgs) hello; }` uses the attribute name (usually `pname`) and with the type you proposed we cannot rename this to a more stable name.

Reply:
> as `{ inherit (pkgs) hello; }` uses the attribute name (usually `pname`)

Why would it do that? This is `{ hello = pkgs.hello; }` where the first `hello` is provided by the module author. That's as stable as it gets. (I'd recommend against `inherit` syntax here, but that's besides the point)

> we cannot rename

We don't need to rename anything. Let the module author write the correct name. That's the only way we don't accidentally use bad data.

Reply:
It's not the module author, it's the attribute in `pkgs`.

The point is that without renaming, we cannot get a stable interface for overriding, since the name in the final attrset will depend on how the package was added, which means that stuff breaks when people migrate from the list syntax to the attrset syntax.

Reply:
Ok, I see where you are coming from. You don't want to enable overriding of packages added through the list syntax.
I see arguments for both approaches, but I don't care enough to have this PR blocked on this, so I implemented it in the way you suggested now.

Should we then also deprecate the list syntax and print a warning when it's used, so that we can migrate to the new syntax that allows for overriding?

Reply:
We could deprecate it when 23.05 is EOL, so that third party modules that target the supported releases aren't stuck between a warning and an error. This can be achieved with `warnIf (lib.isInOldestRelease 2311)`.